byron-proxy/src/lib/Ouroboros/Byron/Proxy/ChainSync/Server.hs

```diff
@@ -75,91 +75,104 @@ type PollT m = forall f . (Monad (f m), MonadTrans f) => Poll (f m)
   -> DB m
   -> Maybe Point
   -> ResourceT m Point
-pollForTipChange err poll db mPoint = poll takePoint (lift (DB.readTip db))
+pollForTipChange epochSlots err poll db mPoint = poll takePoint (lift (DB.readTip db))
   where
   takePoint :: DB.Tip -> ResourceT m (Maybe Point)
   takePoint tip = case mPoint of
     -- There's no existing point, so we'll take the tip unless the DB is empty.
     Nothing -> case tip of
       DB.TipGenesis -> pure Nothing
       DB.TipEBB slot hash _ -> pure $ Just (Point slot hash)
-      DB.TipBlock slot bytes -> case DB.decodeFull cslBlockDecoder (Lazy.fromStrict bytes) of
+      DB.TipBlock slot bytes -> case decodeBlock epochSlots (Lazy.fromStrict bytes) of
         Left cborError -> lift $ err cborError
-        Right blk      -> pure $ Just $ Point slot (CSL.headerHash blk)
+        Right ablk     -> case Binary.unAnnotated ablk of
+          Cardano.ABOBBlock blk  -> pure $ Just $ Point slot (Cardano.blockHashAnnotated blk)
+          -- FIXME deal with this better.
+          Cardano.ABOBBoundary _ -> error "Corrupt DB: EBB where block expected"
     -- Must wait for a tip better than the given point.
     -- We assume (since there are no forks by assumption) that if the new
     -- tip is not equal to this point, then it's better.
     Just point -> do
-      point' <- lift $ pickBetterTip err point tip
+      point' <- lift $ pickBetterTip epochSlots err point tip
       pure $ if point == point'
              then Nothing
              else Just point'
 
 pickBetterTip
   :: ( Applicative m )
-  => (forall x . CBOR.DeserialiseFailure -> m x)
+  => Cardano.EpochSlots
+  -> (forall x . Binary.DecoderError -> m x)
   -> Point
   -> DB.Tip
   -> m Point
 -- FIXME this case is actually an error. If we have a tip point but the
 -- DB is empty then something went horribly wrong.
-pickBetterTip _ point DB.TipGenesis = pure point
-pickBetterTip _ point (DB.TipEBB slot hash _) =
+pickBetterTip _ _ point DB.TipGenesis = pure point
+pickBetterTip _ _ point (DB.TipEBB slot hash _) =
   if pointSlot point < slot
   then pure $ Point slot hash
   else pure point
 -- Careful: the `point` could be for an EBB, in which case the slots are the
 -- same but we want to pick the block.
 -- So just use a non-strict equality. Since we're only using immutable DB with
 -- no forks, it is correct.
-pickBetterTip err point (DB.TipBlock slot bytes) =
+pickBetterTip epochSlots err point (DB.TipBlock slot bytes) =
   if pointSlot point <= slot
-  then case DB.decodeFull cslBlockDecoder (Lazy.fromStrict bytes) of
+  then case decodeBlock epochSlots (Lazy.fromStrict bytes) of
     Left cborError -> err cborError
-    Right blk      -> pure $ Point slot (CSL.headerHash blk)
+    Right ablk     -> case Binary.unAnnotated ablk of
+      Cardano.ABOBBlock blk  -> pure $ Point slot (Cardano.blockHashAnnotated blk)
+      -- FIXME deal with this better.
+      Cardano.ABOBBoundary _ -> error "Corrupt DB: EBB where block expected"
```

Comment:
I could change the `DB` types so that it decodes the block or boundary and gives an error if there's a block in an EBB slot or an EBB in a block slot. It would mean the caller doesn't have to always do this annoying check, but at the price of more decoding work.